### PR TITLE
implement vertexai streaming

### LINKED ISSUE #344: [Go] streaming for vertexai plugin
Implement streaming generation for the vertexai plugin.

### PR DESCRIPTION
**Pull request by randall77**
Fixes #344

### CODE REVIEW

File: go/go.sum [hunk 1]
```diff
@@ -5,8 +5,12 @@ cloud.google.com/go/ai v0.5.0 h1:x8s4rDn5t9OVZvBCgtr5bZTH5X0O7JdE6zYo+O+MpRw=
 cloud.google.com/go/ai v0.5.0/go.mod h1:96VBphk70e0zdXZrbtgPuKYRZsQ3UktSUXhuojwiKA8=
 cloud.google.com/go/aiplatform v1.66.0 h1:bbFYY4JInclG10czRFUYj2rjD+obhh3Gi9zVlyoMgEc=
```
Comment (ianlancetaylor) [inline go/go.sum:6]: Run `go mod tidy`.

File: go/plugins/vertexai/vertexai.go [hunk 2]
```diff
@@ -213,13 +211,77 @@ func (g *generator) generate(ctx context.Context, input *ai.GenerateRequest, cb
...
 
-	r := translateResponse(resp)
-	r.Request = input
+	// Streaming version.
+	iter := cs.SendMessageStream(ctx, parts...)
+	r := &ai.GenerateResponse{Request: input, Candidates: make([]*ai.Candidate, input.Candidates)}
+	for {
+		chunk, err := iter.Next()
+		if err != nil {
+			if err.Error() == "no more items in iterator" {
```
Comment (ianlancetaylor) [inline go/plugins/vertexai/vertexai.go:231]: My understanding is that you can write

    if err == iterator.Done

where iterator is google.golang.org/api/iterator/.

https://github.com/googleapis/google-cloud-go/wiki/Iterator-Guidelines

**Review by ianlancetaylor — APPROVED** [review]

File: go/plugins/vertexai/vertexai.go (hunk 2 above)
Comment (jba) [reply go/plugins/vertexai/vertexai.go:231]: Correct.

File: go/plugins/vertexai/vertexai.go [hunk 3]
```diff
@@ -213,13 +211,77 @@ func (g *generator) generate(ctx context.Context, input *ai.GenerateRequest, cb
...
+				Index:   tc.Index,
+			})
+			if err != nil {
+				return nil, err
+			}
+			// Save candidate in full response structure.
+			if old := r.Candidates[tc.Index]; old == nil {
+				r.Candidates[tc.Index] = tc
+			} else {
+				// Need to merge two "parts" of a candidate.
```
Comment (jba) [inline go/plugins/vertexai/vertexai.go:253]: Would it help if vertexai/genai.joinResponses was exported?

Comment (randall77) [reply go/plugins/vertexai/vertexai.go:253]: Yes, probably would. I think the general state of streaming documentation is pretty poor, so it's hard to know for sure.

File: go/plugins/vertexai/vertexai.go (hunk 2 above)
Comment (randall77) [reply go/plugins/vertexai/vertexai.go:231]: I tried that. I couldn't get it to work. But of course, now it works. Maybe that was fixed somewhere between 0.7.1 and 0.10.0.
And of course, it isn't documented how you should do it.

File: go/go.sum (hunk 1 above)
Comment (randall77) [reply go/go.sum:6]: Done.

File: go/plugins/vertexai/vertexai.go (hunk 3 above)
Comment (randall77) [reply go/plugins/vertexai/vertexai.go:254]: That's a good reference though, I can just copy what they do.

Comment (randall77) [reply go/plugins/vertexai/vertexai.go:254]: They just need to export the merged field somehow, they are already calling it. I think you did the same for googleai?
It's almost accessible, I can get the content field from the chat history. But the rest is inaccessible.

Comment (jba) [reply go/plugins/vertexai/vertexai.go:254]: See https://github.com/googleapis/google-cloud-go/pull/10364

Comment (randall77) [reply go/plugins/vertexai/vertexai.go]: Using the newly exported api.

**Pull request merged**